staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go

```diff
@@ -225,10 +255,72 @@ func (r *proxyHandler) updateAPIService(apiService *apiregistrationv1api.APIServ
 		serviceNamespace: apiService.Spec.Service.Namespace,
 		servicePort:      *apiService.Spec.Service.Port,
 		serviceAvailable: apiregistrationv1apihelper.IsAPIServiceConditionTrue(apiService, apiregistrationv1api.Available),
+		defaultSNI:       transportConfig.TLS.ServerName,
+	}
+
+	r.updateHandlingInfo(&newInfo)
+}
+
+func (r *proxyHandler) updateHandlingInfo(handlingInfo *proxyHandlingInfo) {
+	handlingInfo.proxyRoundTripper, handlingInfo.transportBuildingError = transport.New(handlingInfo.transportConfig)
+	if handlingInfo.transportBuildingError != nil {
+		klog.Warning(handlingInfo.transportBuildingError.Error())
+	}
+	r.handlingInfo.Store(*handlingInfo)
+}
+
+func newFallbackRoundTripper(proxyHandler *proxyHandler, roundTripper http.RoundTripper) *fallbackRoundTripper {
+	return &fallbackRoundTripper{
+		proxyHandler:         proxyHandler,
+		originalRoundTripper: roundTripper,
+	}
+}
+
+var _ utilnet.RoundTripperWrapper = &fallbackRoundTripper{}
+
+type fallbackRoundTripper struct {
+	proxyHandler         *proxyHandler
+	originalRoundTripper http.RoundTripper
+}
+
+func (f *fallbackRoundTripper) WrappedRoundTripper() http.RoundTripper {
+	return f.originalRoundTripper
+}
+
+func (f *fallbackRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	value := f.proxyHandler.handlingInfo.Load()
+	if value == nil {
+		return f.originalRoundTripper.RoundTrip(req)
+	}
+	handlingInfo := value.(proxyHandlingInfo)
+
+	// return early if the request is already using the default SNI
+	if handlingInfo.transportConfig == nil || handlingInfo.transportConfig.TLS.ServerName == handlingInfo.defaultSNI {
+		return f.originalRoundTripper.RoundTrip(req)
 	}
-	newInfo.proxyRoundTripper, newInfo.transportBuildingError = transport.New(newInfo.transportConfig)
-	if newInfo.transportBuildingError != nil {
-		klog.Warning(newInfo.transportBuildingError.Error())
+
+	// clone the request for use with retry
+	clone := utilnet.CloneRequest(req)
+
+	resp, err := f.originalRoundTripper.RoundTrip(req)
+	// return error if the error is not an x509 HostnameError
+	if err != nil && !errors.As(err, &x509.HostnameError{}) {
+		return resp, err
 	}
-	r.handlingInfo.Store(newInfo)
+
+	if resp != nil {
```

Comment:
Check the err for nil, not the resp.

Reply:
updated in [2f646a9](https://github.com/kubernetes/kubernetes/pull/123808/commits/2f646a9df7b788f7e297b220711848a68b91c584)